### PR TITLE
Add ruff and fix initial issues

### LINKED ISSUE #158: Migrate to ruff
[ruff](https://github.com/astral-sh/ruff) is a new linter for Python that covers rules from multiple existing standards.

May be worth switching.

### PR DESCRIPTION
**Pull request by oerc0122**
Migrate to ruff with many linters enabled. Remaining linters to be sorted in later PRs. 

Fixes #158

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -43,3 +44,57 @@ Repository="https://github.com/oerc0122/castep_outputs.git"
...
+    # "D",   # Pydocstyle
+    "UP",  # Pyupgrade
+]
+ignore = [
+    "PLR0913",  # Too many arguments
+    "PLR0912",  # Too many branches
+    "PLR0915",  # Too many statements
+    "PLW2901",  # For loop variable overwritten
+    "PLR2004",  # Magic constant value
+    "E741",     # Ambiguous variable name
```
Comment (ajjackson) [inline pyproject.toml:93]: This one only blocks variables named "I", "O" or "l", that seems pretty healthy!

**Review by ajjackson — COMMENTED** [review]
Looks mostly harmless. Most likely "approve" but first a few comments for discussion.

File: castep_outputs/cli/castep_outputs_main.py [hunk 2]
```diff
@@ -49,10 +50,9 @@ def parse_single(in_file: Union[str, Path, TextIO],
         data = normalise(data, {tuple: list})
 
     if testing:
-        if isinstance(data, list):
-            data = [flatten_dict(run) for run in data]
-        else:
-            data = flatten_dict(data)
+        data = ([flatten_dict(run) for run in data]
+                if isinstance(data, Sequence) else
+                flatten_dict(data))
```
[the remark this answers is not in the document]
Comment (oerc0122) [reply castep_outputs/cli/castep_outputs_main.py:55]: Yeah. Ruff complains about `if x: var = a; else: var =b`. I may just turn that off.

The `testing` here is whether the data are being processed for a `testcode` test within CASTEP, so it filters and restructures data, not internal testing of castep_outputs.

File: castep_outputs/parsers/castep_file_parser.py [hunk 3]
```diff
@@ -1016,7 +1017,7 @@ def parse_castep_file(castep_file_in: TextIO,
                 continue
 
             if not (match := re.search(REs.MINIMISERS_RE, line)):
-                raise IOError("Invalid Geom block")
+                raise OSError("Invalid Geom block")
```
[the remark this answers is not in the document]
Comment (oerc0122) [reply castep_outputs/parsers/castep_file_parser.py:1020]: True, will fix.

File: pyproject.toml [hunk 4]
```diff
@@ -43,3 +44,57 @@ Repository="https://github.com/oerc0122/castep_outputs.git"
...
+    "PERF",# Perflint
+    # "D",   # Pydocstyle
+    "UP",  # Pyupgrade
+]
+ignore = [
+    "PLR0913",  # Too many arguments
+    "PLR0912",  # Too many branches
+    "PLR0915",  # Too many statements
+    "PLW2901",  # For loop variable overwritten
+    "PLR2004",  # Magic constant value
```
[the remark this answers is not in the document]
Comment (oerc0122) [reply pyproject.toml:92]: Where I'm checking for lines of a specific length it flags. The reason for the magic numbers is obvious and I think peppering the code with disabling flags is worse.
```
if split and len(parsed[0]) == 6
```

File: pyproject.toml (hunk 1 above)
Comment (oerc0122) [reply pyproject.toml:93]: `l` is a genuine useful name in `PSPot` information. I suppose I could disable it for that line.
```python
class PSPotTableInfo(TypedDict, total=False):
    rc: float
    beta: Union[float, str]
    e: float
    j: int
    l: int
    norm: int
    scheme: Literal["2b", "es", "ev", "fh", "pn", "pv", "qb", "qc", "tm"]
```

File: pyproject.toml (hunk 4 above)
[the remark this answers is not in the document]
Comment (ajjackson) [reply pyproject.toml:92]: Ok seems reasonable. Parsers have different needs from "normal" code.

**Review by ajjackson — APPROVED** [review]
Ok I think that's all for now 😄

**Pull request merged**